### DISCUSSION TITLE
How to build equivariant NN with coefficients of spherical harmonics as inputs?

### DISCUSSION THREAD

**Discussion by ray38** [e3nn/e3nn#417] [Q&A]
Dear developers,

  First thank you for this great project! I would love to use e3nn to build a very simple equivariant model with inputs that are already projected onto the basis of real spherical harmonics, but couldn't figure out how after several trials, so I would really appreciate some experts' help.

  so the idea is simple, I would like to build a model that predicts vector (say it's the force experienced by a particle), from a set of coefficients of real spherical harmonics. i.e. the environment of the particle is already projected onto the basis of real spherical harmonics of different degrees and different orders. I would like to just build a equivariant NN model that, when the environment is rotated (hence the input coefficients will change according to the Wigner D matrix), the prediction (force) would also change accordingly. Since it does not involve any graph operation or convolutions, but just a fully connected neural network, I figure it's probably a trivial task for e3nn.

  After some reading, I had the impression that the coefficients are the irreps so I just built the following simple model to test:
  
```
class FCTPModel(torch.nn.Module):
    def __init__(self, irreps_in):
        super().__init__()
        self.irreps_in = irreps_in
        self.irreps_out = o3.Irreps("1o")
        # A simple equivariant layer
        self.tp = FullyConnectedTensorProduct(
            irreps_in1=self.irreps_in,
            irreps_in2=self.irreps_in,
            irreps_out=self.irreps_out,
        )
    def forward(self, x):
        return self.tp(x, x)

model = FCTPmodel(o3.Irreps(["0e", "1e", "2o", "3e", "4o"]))
```

and the inputs are provided as a long 1d vector, in this case with 25 values, concatenating the coefficients of real spherical harmonics: [Y_0,0, Y_1,-1, Y_1,0, Y_1,1, Y_2,-2, Y_2,-1, Y_2,0, ....]

  However, when testing the model, when I rotate the inputs, the outputs do not rotate accordingly at all, so I must have some deep misunderstanding of the theory... could you help me understand why is the model not equivariant under rotation here? Also, the model here is based on a simple tensor product of the input, which is probably not the ideal way of building a deep equivariant neural network. is e3nn.nn.FullyConnectedNet the right model to use instead?

  This is probably a very naive problem, but if you could point me to a solution, or some resources it would be really appreciated.

Thanks again!

**Comment by mariogeiger** [comment]
Try more because everything looks good

**Comment by ray38** [comment]
Thanks a lot @mariogeiger for the quick response!

just so that we are on the same page, could you help me with a few quick questions?
- The coefficients of real harmonics should be used (not the complex ones), and the sequential order of the degrees (-l to +l) for each order is correct, right?
- Should I expect the model to be completely equivariant even without training? 
- To make deeper networks, should I use FullyConnectedNet instead, or it's more preferable to just add more TP layers to the current model?

Thanks again!

**Reply by mariogeiger** [reply]
- yes real. But check that you use the same reps as e3nn
- it's always equivariant, for any parameters 
- there is many ways to make it deep, you need to have some activation functions
